phoenix-core/src/main/java/org/apache/phoenix/iterate/DeferredByteBufferSegmentQueue.java

```diff
@@ -0,0 +1,123 @@
+package org.apache.phoenix.iterate;
+
+import org.apache.commons.io.output.DeferredFileOutputStream;
+import org.apache.phoenix.memory.MemoryManager;
+import org.apache.phoenix.memory.MemoryManager.MemoryChunk;
+
+import java.io.*;
+import java.util.*;
+
+public abstract class DeferredByteBufferSegmentQueue<T> extends BufferSegmentQueue<T> {
+
+    final MemoryChunk chunk;
+
+    public DeferredByteBufferSegmentQueue(int index, int thresholdBytes,
+                                          boolean hasMaxQueueSize, MemoryManager memoryManager) {
+        super(index, thresholdBytes, hasMaxQueueSize);
+        chunk = memoryManager.allocate(thresholdBytes);
+    }
+
+    abstract protected void writeToBuffer(OutputStream outputStream, T e);
+    abstract protected T readFromBuffer(DataInput dataInput);
+
+
+    @Override
+    protected SegmentQueueFileIterator createSegmentQueueFileIterator(SegmentQueueFileIterator iterator){
+        return new DeferredSegmentQueueFileIterator(iterator);
+    }
+
+    @Override
+    protected SegmentQueueFileIterator createSegmentQueueFileIterator(){
+        return new DeferredSegmentQueueFileIterator();
+    }
+
+    @Override
+    protected void flush(T entry) throws IOException {
+        Queue<T> inMemQueue = getInMemoryQueue();
+        int resultSize = sizeOf(entry);
+        maxResultSize = Math.max(maxResultSize, resultSize);
+        totalResultSize = hasMaxQueueSize ? maxResultSize * inMemQueue.size() : (totalResultSize + resultSize);
+        if (totalResultSize >= thresholdBytes) {
+            this.file = File.createTempFile(UUID.randomUUID().toString(), null);
+
+            DeferredFileOutputStream spoolTo = new DeferredFileOutputStream(thresholdBytes, file) {
+                @Override
+                protected void thresholdReached() throws IOException {
+                    try {
+                        super.thresholdReached();
+                    } finally {
+                            chunk.close();
```

Comment:
There's an indent problem here. You might not need these lines after the suggested change, but would you mind double checking other places for code indent?